codecarbon/external/hardware.py

```diff
@@ -52,46 +52,64 @@ def start(self) -> None:
 
 @dataclass
 class GPU(BaseHardware):
-    num_gpus: int
     gpu_ids: Optional[List]
 
     def __repr__(self) -> str:
         return super().__repr__() + " ({})".format(
-            ", ".join([d["name"] for d in get_gpu_details()])
+            ", ".join([d["name"] for d in self.devices.get_gpu_details()])
         )
 
-    def _get_power_for_gpus(self, gpu_ids: Iterable[int]) -> Power:
-        """
-        Get total power consumed by specific GPUs identified by `gpu_ids`
-        :param gpu_ids:
-        :return: power in kW
-        """
-        all_gpu_details: List[Dict] = get_gpu_details()
-        return Power.from_milli_watts(
+    def __post_init__(self):
+        self.devices = AllGPUDevices()
+        self.num_gpus = self.devices.device_count
+
+    def measure_power_and_energy(self, last_duration: float) -> Tuple[Power, Energy]:
+        gpu_ids = self._get_gpu_ids()
+        all_gpu_details: List[Dict] = self.devices.get_delta(
+            Time.from_seconds(last_duration)
+        )
+        # We get the total energy and power of only the ones in gpu_ids
+        total_energy = Energy.from_energy(
             sum(
                 [
-                    gpu_details["power_usage"]
+                    gpu_details["delta_energy_consumption"].kWh
                     for idx, gpu_details in enumerate(all_gpu_details)
                     if idx in gpu_ids
                 ]
             )
         )
+        total_power = Power(
+            sum(
+                [
+                    gpu_details["power_usage"].kW
+                    for idx, gpu_details in enumerate(all_gpu_details)
+                    if idx in gpu_ids
+                ]
+            )
+        )
+        return total_power, total_energy
 
-    def total_power(self) -> Power:
+    def _get_gpu_ids(self) -> Iterable[int]:
+        """
+        Get the Ids of the GPUs that we would like to monitor
+        :return: list of ids
+        """
         if self.gpu_ids is not None:
             gpu_ids = self.gpu_ids
             assert set(gpu_ids).issubset(
                 set(range(self.num_gpus))
             ), f"Unknown GPU ids {gpu_ids}"
         else:
             gpu_ids = set(range(self.num_gpus))
+        return gpu_ids
 
-        gpu_power = self._get_power_for_gpus(gpu_ids=gpu_ids)
-        return gpu_power
+    def total_power(self) -> Power:
+        # FIXME: do we still need this total_power?
```

Comment:
@benoit-cty maybe we do not need the total power of the gpu anymore? Now the power is taken from converting the energy by calculating the delta. (see `measure_power_and_energy` function)